src/server/replicaton_buffer2.rs

```diff
@@ -0,0 +1,328 @@
+    /// See also [`Self::cancel_entity_data`].
+    pub(super) fn end_entity_data(
+        &mut self,
+        buffer: &mut ReplicationBuffer,
+        entity: Entity,
+        data_len: u16,
+    ) -> bincode::Result<()> {
+        let begin = buffer.position();
+        serialize_entity(&mut buffer.cursor, entity)?;
+        bincode::serialize_into(&mut buffer.cursor, &data_len)?;
+        let end = buffer.position();
+
+        self.ranges[self.entity_data_index] = begin..end;
+
+        Ok(())
+    }
+
+    /// Removes last preallocated range that was created for an entity data.
+    ///
+    /// Should be called only after [`Self::start_entity_data`].
+    /// See also [`Self::end_entity_data`].
+    pub(super) fn cancel_entity_data(&mut self) {
+        debug_assert_eq!(self.entity_data_index, self.ranges.len() - 1);
+
+        self.ranges.pop();
+    }
+
+    /// Serializes replicon tick into the buffer or reuses previously written range if exists.
+    pub(super) fn write_tick(
+        &mut self,
+        buffer: &mut ReplicationBuffer,
+        replicon_tick: RepliconTick,
+    ) -> bincode::Result<()> {
+        if let Some(range) = buffer.current_range() {
+            self.extend_or_push(range);
+        } else {
+            let begin = buffer.start_range();
+            bincode::serialize_into(&mut buffer.cursor, &replicon_tick)?;
+            let end = buffer.position();
+            self.ranges.push(begin..end);
+        }
+
+        Ok(())
+    }
+
+    /// Serializes entity to entity mapping into the buffer or reuses previously written range if exists.
+    pub(super) fn write_client_mapping(
+        &mut self,
+        buffer: &mut ReplicationBuffer,
+        mapping: &ClientMapping,
+    ) -> bincode::Result<()> {
+        if let Some(range) = buffer.current_range() {
+            self.extend_or_push(range);
+        } else {
+            let begin = buffer.start_range();
+            serialize_entity(&mut buffer.cursor, mapping.server_entity)?;
+            serialize_entity(&mut buffer.cursor, mapping.client_entity)?;
+            let end = buffer.position();
+            self.ranges.push(begin..end);
+        }
+
+        Ok(())
+    }
+
+    /// Serializes entity into the buffer or reuses previously written range if exists.
+    pub(super) fn write_entity(
+        &mut self,
+        buffer: &mut ReplicationBuffer,
+        entity: Entity,
+    ) -> bincode::Result<()> {
+        if let Some(range) = buffer.current_range() {
+            self.extend_or_push(range);
+        } else {
+            let begin = buffer.start_range();
+            serialize_entity(&mut buffer.cursor, entity)?;
+            let end = buffer.position();
+            self.ranges.push(begin..end);
+        }
+
+        Ok(())
+    }
+
+    /// Serializes component and its replication ID into the buffer or reuses previously written range if exists.
+    pub(super) fn write_component(
```

Comment:
The range will be extended if the next serialized component will be sent to client. So it will be per entity if client will receive all components that continuously serialized.

Reply:
To clarify: in theory the range could cover everything until the array and entity data lengths because they written in the end, but use older range that I preallocated at the beginning of array or entity data.